File: src/test/java/no/nav/sosialhjelp/soknad/business/batch/AvbrytAutomatiskSchedulerTest.java
package no.nav.sosialhjelp.soknad.business.batch;

import no.nav.sosialhjelp.soknad.business.db.config.DbTestConfig;
import no.nav.sosialhjelp.soknad.business.db.repositories.soknadmetadata.BatchSoknadMetadataRepository;
import no.nav.sosialhjelp.soknad.business.db.repositories.soknadmetadata.SoknadMetadataRepository;
import no.nav.sosialhjelp.soknad.business.db.repositories.soknadunderarbeid.BatchSoknadUnderArbeidRepository;
import no.nav.sosialhjelp.soknad.business.domain.SoknadMetadata;
import no.nav.sosialhjelp.soknad.client.leaderelection.LeaderElection;
import no.nav.sosialhjelp.soknad.domain.SoknadMetadataInnsendingStatus;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeidStatus;
import no.nav.sosialhjelp.soknad.domain.model.kravdialoginformasjon.SoknadType;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.test.context.ContextConfiguration;

import java.time.LocalDateTime;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
@ContextConfiguration(classes = {DbTestConfig.class})
class AvbrytAutomatiskSchedulerTest {

    private static final String EIER = "11111111111";
    private static final String BEHANDLINGS_ID = "1100AAAAA";
    private static final int DAGER_GAMMEL_SOKNAD = 14;

    @Mock
    private LeaderElection leaderElection;
    @Mock
    private BatchSoknadUnderArbeidRepository batchSoknadUnderArbeidRepository;
    @Mock
    private SoknadMetadataRepository soknadMetadataRepository;
    @Mock
    private BatchSoknadMetadataRepository batchSoknadMetadataRepository;

    private AvbrytAutomatiskSheduler scheduler;

    @BeforeEach
    public void setup() {
        scheduler = new AvbrytAutomatiskSheduler(leaderElection, soknadMetadataRepository, batchSoknadMetadataRepository, batchSoknadUnderArbeidRepository, true);

        when(leaderElection.isLeader()).thenReturn(true);
    }

    @Test
    void avbrytAutomatiskOgSlettGamleSoknader() {
        SoknadMetadata soknadMetadata = soknadMetadata(BEHANDLINGS_ID, SoknadMetadataInnsendingStatus.UNDER_ARBEID, DAGER_GAMMEL_SOKNAD + 1);
        SoknadUnderArbeid soknadUnderArbeid = new SoknadUnderArbeid()
                .withSoknadId(1L)
                .withEier(EIER)
                .withBehandlingsId(BEHANDLINGS_ID)
                .withStatus(SoknadUnderArbeidStatus.UNDER_ARBEID);

        when(batchSoknadMetadataRepository.hentForBatch(DAGER_GAMMEL_SOKNAD)).thenReturn(Optional.of(soknadMetadata)).thenReturn(Optional.empty());
        when(batchSoknadUnderArbeidRepository.hentSoknadUnderArbeidIdFromBehandlingsIdOptional(BEHANDLINGS_ID)).thenReturn(Optional.of(soknadUnderArbeid.getSoknadId()));

        scheduler.avbrytGamleSoknader();

        ArgumentCaptor<SoknadMetadata> argument = ArgumentCaptor.forClass(SoknadMetadata.class);
        verify(soknadMetadataRepository).oppdater(argument.capture());
        SoknadMetadata oppdatertSoknadMetadata = argument.getValue();

        assertThat(oppdatertSoknadMetadata.status).isEqualTo(SoknadMetadataInnsendingStatus.AVBRUTT_AUTOMATISK);
        verify(batchSoknadUnderArbeidRepository).slettSoknad(anyLong());
    }

    private SoknadMetadata soknadMetadata(String behandlingsId, SoknadMetadataInnsendingStatus status, int dagerSiden) {
        SoknadMetadata meta = new SoknadMetadata();
        meta.id = soknadMetadataRepository.hentNesteId();
        meta.behandlingsId = behandlingsId;
        meta.fnr = EIER;
        meta.type = SoknadType.SEND_SOKNAD_KOMMUNAL;
        meta.skjema = "";
        meta.status = status;
        meta.innsendtDato = LocalDateTime.now().minusDays(dagerSiden);
        meta.opprettetDato = LocalDateTime.now().minusDays(dagerSiden);
        meta.sistEndretDato = LocalDateTime.now().minusDays(dagerSiden);

        return meta;
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/scheduled/AvbrytAutomatiskSchedulerTest.kt
package no.nav.sosialhjelp.soknad.scheduled

import io.mockk.every
import io.mockk.just
import io.mockk.mockk
import io.mockk.runs
import io.mockk.slot
import io.mockk.verify
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.BatchSoknadMetadataRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadata
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadataInnsendingStatus
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadataInnsendingStatus.UNDER_ARBEID
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadataRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadataType
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.BatchSoknadUnderArbeidRepository
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.db.repositories.soknadunderarbeid.SoknadUnderArbeidStatus
import no.nav.sosialhjelp.soknad.innsending.SoknadService
import no.nav.sosialhjelp.soknad.scheduled.leaderelection.LeaderElection
import no.nav.sosialhjelp.soknad.vedlegg.fiks.MellomlagringService
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

internal class AvbrytAutomatiskSchedulerTest {
    private val leaderElection: LeaderElection = mockk()
    private val batchSoknadUnderArbeidRepository: BatchSoknadUnderArbeidRepository = mockk()
    private val soknadMetadataRepository: SoknadMetadataRepository = mockk()
    private val batchSoknadMetadataRepository: BatchSoknadMetadataRepository = mockk()
    private val mellomlagringService: MellomlagringService = mockk()

    private val scheduler = AvbrytAutomatiskScheduler(
        batchEnabled = true,
        schedulerDisabled = false,
        leaderElection,
        soknadMetadataRepository,
        batchSoknadMetadataRepository,
        batchSoknadUnderArbeidRepository,
        mellomlagringService,
    )

    @BeforeEach
    fun setup() {
        every { leaderElection.isLeader() } returns true
        every { soknadMetadataRepository.hentNesteId() } returns 123L
    }

    @Test
    fun avbrytAutomatiskOgSlettGamleSoknader() {
        val soknadMetadata = soknadMetadata(BEHANDLINGS_ID, UNDER_ARBEID, DAGER_GAMMEL_SOKNAD + 1)
        val soknadUnderArbeid = SoknadUnderArbeid(
            soknadId = 1L,
            versjon = 1L,
            behandlingsId = BEHANDLINGS_ID,
            tilknyttetBehandlingsId = null,
            eier = "11111111111",
            jsonInternalSoknad = null,
            status = SoknadUnderArbeidStatus.UNDER_ARBEID,
            opprettetDato = LocalDateTime.now(),
            sistEndretDato = LocalDateTime.now()
        )

        every {
            batchSoknadMetadataRepository.hentForBatch(DAGER_GAMMEL_SOKNAD)
        } returns soknadMetadata andThen null

        every {
            batchSoknadUnderArbeidRepository.hentSoknadUnderArbeid(BEHANDLINGS_ID)
        } returns soknadUnderArbeid

        val soknadMetadataSlot = slot<SoknadMetadata>()
        every { soknadMetadataRepository.oppdater(capture(soknadMetadataSlot)) } just runs

        every { mellomlagringService.erMellomlagringEnabledOgSoknadSkalSendesMedDigisosApi(soknadUnderArbeid) } returns false

        scheduler.avbrytGamleSoknader()

        verify { soknadMetadataRepository.oppdater(soknadMetadataSlot.captured) }
        val oppdatertSoknadMetadata = soknadMetadataSlot.captured
        assertThat(oppdatertSoknadMetadata.status)
            .isEqualTo(SoknadMetadataInnsendingStatus.AVBRUTT_AUTOMATISK)
        verify { batchSoknadUnderArbeidRepository.slettSoknad(any()) }
        verify(exactly = 0) { mellomlagringService.deleteAllVedlegg(any()) }
    }

    @Test
    fun avbrytAutomatiskOgSlettGamleSoknaderMedMellomlagredeVedlegg() {
        val soknadMetadata = soknadMetadata(BEHANDLINGS_ID, UNDER_ARBEID, DAGER_GAMMEL_SOKNAD + 1)
        val soknadUnderArbeid = SoknadUnderArbeid(
            soknadId = 1L,
            versjon = 1L,
            behandlingsId = BEHANDLINGS_ID,
            tilknyttetBehandlingsId = null,
            eier = "11111111111",
            jsonInternalSoknad = SoknadService.createEmptyJsonInternalSoknad("11111111111"),
            status = SoknadUnderArbeidStatus.UNDER_ARBEID,
            opprettetDato = LocalDateTime.now(),
            sistEndretDato = LocalDateTime.now()
        )

        soknadUnderArbeid.jsonInternalSoknad?.soknad?.mottaker?.kommunenummer = "1234"

        every {
            batchSoknadMetadataRepository.hentForBatch(DAGER_GAMMEL_SOKNAD)
        } returns soknadMetadata andThen null
        every {
            batchSoknadUnderArbeidRepository.hentSoknadUnderArbeid(BEHANDLINGS_ID)
        } returns soknadUnderArbeid

        val soknadMetadataSlot = slot<SoknadMetadata>()
        every { soknadMetadataRepository.oppdater(capture(soknadMetadataSlot)) } just runs

        every { mellomlagringService.erMellomlagringEnabledOgSoknadSkalSendesMedDigisosApi(soknadUnderArbeid) } returns true
        every { mellomlagringService.deleteAllVedlegg(any()) } just runs

        scheduler.avbrytGamleSoknader()

        verify { soknadMetadataRepository.oppdater(soknadMetadataSlot.captured) }
        val oppdatertSoknadMetadata = soknadMetadataSlot.captured
        assertThat(oppdatertSoknadMetadata.status)
            .isEqualTo(SoknadMetadataInnsendingStatus.AVBRUTT_AUTOMATISK)
        verify { batchSoknadUnderArbeidRepository.slettSoknad(any()) }
        verify(exactly = 1) { mellomlagringService.deleteAllVedlegg(any()) }
    }

    private fun soknadMetadata(
        behandlingsId: String,
        status: SoknadMetadataInnsendingStatus,
        dagerSiden: Int
    ): SoknadMetadata {
        return SoknadMetadata(
            id = soknadMetadataRepository.hentNesteId(),
            behandlingsId = behandlingsId,
            fnr = EIER,
            type = SoknadMetadataType.SEND_SOKNAD_KOMMUNAL,
            skjema = "",
            status = status,
            innsendtDato = LocalDateTime.now().minusDays(dagerSiden.toLong()),
            opprettetDato = LocalDateTime.now().minusDays(dagerSiden.toLong()),
            sistEndretDato = LocalDateTime.now().minusDays(dagerSiden.toLong())
        )
    }

    companion object {
        private const val EIER = "11111111111"
        private const val BEHANDLINGS_ID = "1100AAAAA"
        private const val DAGER_GAMMEL_SOKNAD = 14
    }
}
